### ISSUE TITLE
Context when validating data

### ISSUE THREAD

**Issue by samuelcolvin** [pydantic/pydantic#1549]
This is discussed at length in #1170 but I'm creating a new issue to make things clearer.

---

Pydantic needs a way of accessing "context" when validating data, serialising data, creating schema.

Might be used via `MyModel.parse_obj(raw_data, context=my_context)`.

Use cases:
* dynamic choices - E.g. when choosing from a select based on a entities you have access to in a db, obviously both the validation and schema for the field should be dynamic - this is very common frustration for me
* dynamic parameters on fields, e.g. max and min
* cases like [this](https://github.com/samuelcolvin/sasstastic/blob/5777b98e328bbdcee99a984f3bc1909e36584e6e/sasstastic/config.py#L77-L88)
* and so many more...

Questions/concerns:
* adding context with `parse_obj`, but I don't see way way to add context with `MyModel(**data)` maybe that's fine
* how do we include context to field definitions? e.g. `foobar: conint(gt=context['...'])` won't work
* how will context work in sub-models?
* will this require breaking changes that need it to go into v2? I hope not.

Another option is to encourage one of the options from #1170 and/or make one of them easier.

Feedback on how this might work and when you would use it very welcome.

**Comment by simkimsia** [comment]
Is it better to split some of the concerns or use cases into separate issues/features?

Meaning to say maybe context will only solve a subset of the concerns here for now

**Comment by samuelcolvin** [comment]
no, this is one feature which needs one discussion.

**Comment by simkimsia** [comment]
Copy that @samuelcolvin

**Comment by anentropic** [comment]
personally right now I'm interested in #1575 use case (auto-adding an attribute to child model instances which links to the parent model instance) and don't care much about context-aware serialization

so this is a comment in support of any solution which makes the former possible

**Comment by michaelurban** [comment]
I'm migrating to pydantic from marshmallow but am blocked on this. Any movement?

**Comment by samuelcolvin** [comment]
No movement is far from trivial, but it's definitely something I want.

Your best bet for now might be to use contextvars.

It would be helpful if you could give feedback on my questions at the top of this page, or show us your intended use case.

**Comment by michaelurban** [comment]
Thanks for the quick reply. I've updated my comment with something similar to my use case.

**Comment by nandoflorestan** [comment]
I am validating the email address of a new user.  Depending on the runtime environment (local, staging, production etc.) we have certain domains blacklisted.  This is a job for a context object (the blocked domains are a configuration that depends on the environment).  The validator function needs to be aware of the context.

I don't know how you're going to adapt the pydantic API to accomplish this -- just wanted to add an example of a common use case to this discussion.  Similar to this use case, often my validators need access to my repository object in order to query the database -- yet another thing that a context variable in a validator function would solve.

The *colander* library solves this by allowing the user to bind a schema instance to certain variables at runtime, e. g.:

    my_schema = SomeSchemaClass().bind(request=request, repository=repository)

...and then the validator function can do this:

    def email_domain_allowed(node, value):
        request = node.bindings["request"]
        ...

So in Colander, the context bindings are a property of the node, rather than a 3rd argument to validator functions.

Maybe the above will help in some way... if not, sorry for the noise.

**Comment by MadPepe** [comment]
Adding context in a clear functionalities and passing it to the nested models in a clean fashion would be a game changer for me. Finger crossed that it will be considered in near future.

**Comment by mjog** [comment]
Dynamic choices is my use case for this, but I typically use pydantic via FastAPI, so the ability to hook it up to FAPI's router and response_model magic would be essential. I'm not sure how either has been implemented, but I assume it uses parse_obj/parse_orm/etc, so passing context via those that might work well.

Note also FAPI's [dependency's injection system](https://fastapi.tiangolo.com/tutorial/dependencies/) is pretty flexible - it might provide a solid basis for the same here.

**Comment by michaeloliverx** [comment]
I had a need for this recently when I was parsing a spreadsheet and wanted to include the cell location in error messages.

**Comment by DrChai** [comment]
for #1575 or other related use cases, a workaround solution is to create a field for the context in child and inject your context data from the validator in the parent model.

```python
class Product(BaseModel):
    id: int
    price: condecimal(decimal_places=2, max_digits=7)
    options: List[Option]
   
    @validator('options', each_item=True)
    def validate_options(cls, value, values):
        allowed_options = get_allowed_opts(values['id'])
        value |= {'allowed_choice': allowed_choice} 
        return value

class Option(BaseModel):
    allowed_choice: Any # your context data for validation
    value: Union[int, float, str]

    @validator('value')
    def check_options(cls, value, values):
        context = values.get('allowed_choice')
        # validating your value with context
```
just remember to exclude the context field before exporting to json or dict

**Comment by samuelcolvin** [comment]
It's already working in pydantic-core and will be fixed in pydantic V2.

See [here](https://pydantic-docs.helpmanual.io/blog/pydantic-v2/) for details on the roadmap.

**Comment by sisp** [comment]
Perhaps a little late since the v2 roadmap has been set, but possibly still relevant/helpful for the future development of Pydantic:

There's a state management library implemented in TypeScript called [`mobx-keystone`](https://github.com/xaviergonz/mobx-keystone/) which makes managing data/state trees very convenient. Among numerous other features, it supports [contexts](https://mobx-keystone.js.org/contexts) in a very elegant, flexible, _and type-safe_ way. Besides injecting context to the call of the validation method (e.g. `<Model>.model_validate_json(<data>, context=<context>)` in the roadmap post), `mobx-keystone`'s context system allows injecting context at any point in the tree, also from within the tree by a model class itself during instantiation. This is extremely powerful in practice.

By the way, the solution proposed for Pydantic v2 is a bit similar to the [dependency injection system of `mobx-state-tree`](https://mobx-state-tree.js.org/concepts/dependency-injection). `mobx-keystone` is basically the next evolution of `mobx-state-tree` written by one of the most active contributors to `mobx-state-tree`.

**Comment by samuelcolvin** [comment]
Thanks should be entirely possible for a wrap validator to alter/inject context which is then accessible by deeper validators.

But I guess there's no test for it at present. Happy to accept a PR to add such a test, either to pydantic or pydantic-core.

**Comment by HappyDingning** [comment]
from typing import Any, List
from pydantic import BaseModel, validator


class BaseModelWithContext(BaseModel):
    context: Any # your context

    @validator('*', each_item=True, pre=True)
    def add_context(cls, value, values, field):
        if isinstance(field.type_, type) and issubclass(field.type_, BaseModelWithContext):
            value['context'] = values['context']
        return value
    

class Pet(BaseModelWithContext):
    name: str


class Member(BaseModelWithContext):
    name: str
    pet: Pet


class Home(BaseModelWithContext):
    id: int
    name: str
    members: List[Member]


home = Home(**{"id": 1, "name": 'my_home', "members": [{"name": 'father', "pet": {"name": "qwe"}}, {"name": 'mother', "pet": {"name": "wer"}}]}, context={"me": "my_name"})
print(home.context)
print(home.members[0].context)
print(home.members[1].context)
print(home.members[0].pet.context)
print(home.members[1].pet.context)

home.context["my_pet"] = "tom"

print(home.context)
print(home.members[0].context)
print(home.members[1].context)
print(home.members[0].pet.context)
print(home.members[1].pet.context)

**Comment by sisp** [comment]
@HappyDingning I think the `context` attribute should be a [private model attribute](https://docs.pydantic.dev/usage/models/#private-model-attributes), so its value doesn't get (de)serialized and it doesn't appear in the schema. WDYT?

**Comment by sisp** [comment]
Or maybe not ... :thinking: Private model attributes cannot be set upon model instantiation. Still, the problem remains how to omit the context from (de)serialization and the schema.

**Comment by HappyDingning** [comment]
> Or maybe not ... 🤔 Private model attributes cannot be set upon model instantiation. Still, the problem remains how to omit the context from (de)serialization and the schema.

```
from typing import Any, List
from pydantic import BaseModel, validator


class BaseModelWithContext(BaseModel):
    context: Any # your context

    @validator('*', each_item=True, pre=True)
    def add_context(cls, value, values, field):
        if isinstance(field.type_, type) and issubclass(field.type_, BaseModelWithContext):
            value['context'] = values['context']
        return value
    
    # newly added code
    def _calculate_keys(self, include, exclude, exclude_unset):  # type: ignore
        if exclude is not None:
            exclude['context'] = ...
        else:
            exclude = {'context': ...}
        return super()._calculate_keys(include=include, exclude=exclude, exclude_unset=exclude_unset)


class Pet(BaseModelWithContext):
    name: str


class Member(BaseModelWithContext):
    name: str
    pet: Pet

    def json(self, **kwargs) -> str:
        print(123)
        return super().json(**kwargs)


class Home(BaseModelWithContext):
    id: int
    name: str
    members: List[Member]


home = Home(**{"id": 1, "name": 'my_home', "members": [{"name": 'father', "pet": {"name": "qwe"}}, {"name": 'mother', "pet": {"name": "wer"}}]}, context={"me": "my_name"})

print(home.json())

# {"id": 1, "name": "my_home", "members": [{"name": "father", "pet": {"name": "qwe"}}, {"name": "mother", "pet": {"name": "wer"}}]}

```

**Comment by dmontagu** [comment]
We have validation context in v2, it can be passed into the `model_validate` method. I have opened a test for this in https://github.com/pydantic/pydantic/pull/5593

I will close this issue once that PR is merged, but please comment if the solution in v2 does not address the desired behavior.

**Comment by jason-wolfe** [comment]
I gave the pydantic v2 alpha a spin, optimistic to use this feature.

Supporting validation context in model_validate but not in constructors makes it very easy to create a broad class of models which can **only** be instantiated via json/str but cannot be created via the type-safe constructor. Is this deliberate? Is it feasible to support a keyword argument on `__init__` to pass context as well?

Examples of functionality likely to encounter this includes:
* conditional validation behavior that requires a runtime toggle
* dynamic choices

Neither of those have sensible default values that can be used like the default multiplier of 1 in #5593. Not passing the parser context indicates a bug and should fail. Non-provided context failing means that the model constructor is unusable.

(meta: Does this discussion belong in the comments here or should that be a new issue?)

**Comment by dmontagu** [comment]
@jason-wolfe — Today, I see two ways to get validation context into `__init__`, both involving overriding the `BaseModel.__init__` method.

The first one, which I like a bit less, is to add a positional argument to the `BaseModel.__init__` signature to accept the context:
```python
from typing import Any

from pydantic_core.core_schema import ValidationInfo

from pydantic import BaseModel, field_validator


class Model(BaseModel):
    x: int

    @field_validator("x")
    @classmethod
    def validate_x(cls, v: int, info: ValidationInfo) -> int:
        return v * info.context.get("multiplier", 1)

    def __init__(__pydantic_self__, __context: dict[str, Any] | None = None, **data: Any) -> None:  # type: ignore
        # `__tracebackhide__` tells pytest and some other tools to omit this function from tracebacks
        __tracebackhide__ = True
        __pydantic_self__.__pydantic_validator__.validate_python(
            data, self_instance=__pydantic_self__, context=__context
        )


context = {"multiplier": 2}
print(Model(context, x=1))
# > x=2
```
This has the downside that it probably won't have nice type-checker integration (at least not without the mypy plugin, and even then, would likely require some changes), which was kind of the whole point of not using `model_validate.

The other approach, which I will admit feels a bit hacky but does work without changes to the signature of `__init__` is to use a `ContextVar`:
```python
from contextlib import contextmanager
from contextvars import ContextVar
from typing import Any

from pydantic_core.core_schema import ValidationInfo

from pydantic import BaseModel, field_validator

_validation_context: ContextVar[dict[str, Any] | None] = ContextVar('_validation_context', default=None)


@contextmanager
def validation_context(context: dict[str, Any] | None) -> None:
    token = _validation_context.set(context)
    try:
        yield
    finally:
        _validation_context.reset(token)


class ContextModel(BaseModel):

    def __init__(__pydantic_self__, __context: dict[str, Any] | None = None, **data: Any) -> None:  # type: ignore
        # `__tracebackhide__` tells pytest and some other tools to omit this function from tracebacks
        __tracebackhide__ = True
        __pydantic_self__.__pydantic_validator__.validate_python(
            data, self_instance=__pydantic_self__, context=_validation_context.get()
        )


class MyModel(ContextModel):
    x: int

    @field_validator("x")
    @classmethod
    def validate_x(cls, v: int, info: ValidationInfo) -> int:
        return v * (info.context or {}).get("multiplier", 1)

# Usage:

print(MyModel(x=1))  # no context, i.e., info.context is None
# > x=1

with validation_context({"multiplier": 2}):
    print(MyModel(x=1))
    # > x=2
```

At any rate, I think you've raised a good point, and if you feel that neither of these approaches are adequate I would suggest you create a new issue describing your issues with these solutions. Ideally if you don't like these, it would be nice if you could specify an API you would prefer.

**Comment by TiansuYu** [comment]
I had issues before, where a lot of my models require data from their parent model (container model) to fulfill some default values. This is my solution for tracking parent model context in a child (therefore all context data from parent model): 
```python
from pydantic import BaseModel


class ParentTrackingModel(BaseModel):
    __parent: BaseModel | None = None

    def model_post_init(self, __context: any) -> None:
        for field_name in self.model_fields:
            child = getattr(self, field_name)
            if isinstance(child, BaseModel):
                child.__parent = self

    @property
    def parent(self):
        return self.__parent


class ChildExample(ParentTrackingModel):
    name: str


class ParentExample(ParentTrackingModel):
    name: str
    child: ChildExample


c = ChildExample(name="Martin")
p = ParentExample(name="Tilda", child=c)

print(c.parent)
```
